LDA ordination with feature loadings computation

theotimeee opened the pull request:
This PR adds two new methods `getLDA` and `addLDA` that perform Latent Dirichlet Allocation.

`getLDA` returns the ordination matrix with the feature loadings matrix stored as attribute.
`addLDA` adds this ordination matrix to `reducedDims` of the TreeSummarizedExperiment.

This PR is related to issue https://github.com/microbiome/mia/issues/596.

Commit:
new functions getLDA and addLDA

Commit:
up

Commit:
add topicmodels package to imports

antagomir commented:
Great! I will be first looking fwd to comments from @TuomasBorman and then let's see if I have anything to add.

antagomir commented:
Conflicts to resolve?

theotimeee committed:
Merge branch 'devel' into feature_loadings_methods

TuomasBorman commented on R/runLDA.R:
```diff
@@ -0,0 +1,75 @@
+#'  object.
+#'   
+#' @param k Integer, number of latent vectors/topics 
+#'   (By default: \code{k = 2}).
```
(Default: \code{2})

TuomasBorman commented on R/runLDA.R:
```diff
@@ -0,0 +1,75 @@
+#' 
+#' @return 
+#' For \code{getLDA}, the ordination matrix with feature loadings matrix
+#'  as attribute \code{'feature_loadings'}.
```
Consider just "loadings"

TuomasBorman commented on R/runLDA.R:
The name of the file should reflect the functions (there is no runLDA function. Other files might be named like run* because names are relics)

TuomasBorman commented on R/runLDA.R:
```diff
@@ -0,0 +1,75 @@
+
+#' @export
+#' @rdname addLDA
+setMethod("addLDA", "SummarizedExperiment",
```
What if user calls this function with SE? --> This does not work. You should update the function to first TreeSE if there is no reducedDim slot (SE does not have one)

TuomasBorman commented on R/runLDA.R:
```diff
@@ -0,0 +1,75 @@
+#' @rdname addLDA
+setMethod("addLDA", "SummarizedExperiment",
+    function(x, k=2, name = "LDA", ...){
+        reducedDim(x, name) <- t(getLDA(x, k, ...))
```
Similarly to internal function ".add_values_to_coldata", we could have ".add_values_to_reduceddim". That could

1. Check if the object is at least SCE. If not, update the object and give warning.
2. Add the result.

That way we could harmonize these functions

TuomasBorman commented on R/runLDA.R:
```diff
@@ -0,0 +1,75 @@
+#' @export
+#' @rdname addLDA
+setMethod("getLDA", "SummarizedExperiment",
+    function(x, k=2, ...){
```
Add assay.type parameter and input checks

TuomasBorman commented on R/runLDA.R:
```diff
@@ -0,0 +1,75 @@
+#' @rdname addLDA
+setMethod("getLDA", "SummarizedExperiment",
+    function(x, k=2, ...){
+        lda_model <- topicmodels::LDA(t(assay(x, "counts")), k)
```
This is another dependency. Because this function is not used that often, we could

1. Add this dependency to suggests
2. Add .require_package("topicmodels") in the beginning of this function.

TuomasBorman commented on R/runLDA.R:
```diff
@@ -0,0 +1,75 @@
+setMethod("getLDA", "SummarizedExperiment",
+    function(x, k=2, ...){
+        lda_model <- topicmodels::LDA(t(assay(x, "counts")), k)
+        df <- as.data.frame(t(assay(x, "counts")))
```
You could do this line first and then just input this variable to line 58 (simplify)

TuomasBorman commented on R/runLDA.R:
```diff
@@ -0,0 +1,75 @@
+#' @export
+#' @rdname addLDA
+setMethod("addLDA", "SummarizedExperiment",
+    function(x, k=2, name = "LDA", ...){
```
k=2 and name = "LDA"

--> k = 2 --> use similar coding style

TuomasBorman reviewed (changes requested):
Looks good. Even though the code is rather short, structure it by adding tests and comments

Commit:
coding style and small modifications

Commit:
Merge branch 'feature_loadings_methods' of https://github.com/microbiome/mia into feature_loadings_methods

Commit:
new internal function add_values_to_reducedDims

Commit:
move topicmodels package to suggests

Commit:
simplify function

TuomasBorman commented:
Couple things to discuss:

1. LDA has features from both ordination and clustering methods, however, it is more clustering method. That should be taken into account in documentation.
2. As this is clustering method, should we add clusters to rowData/colData? In theory, LDA can be applied also to rows, which makes reducedDim not suitable for storing the result. (Instead of ordination methods, addCluster() should be used as an example)
3. And finally, should this be implemented in bluster package (the Bioconductor clustering package)? We could make a PR as we did last year https://github.com/LTLA/bluster/blob/master/R/DmmParam.R

antagomir commented:
Intuitively, reducedDim is a logical place given that this would be usually done for samples. But it could be in principle applied on both rows and columns. The function generates both feature matrix and feature loading matrix, both should be stored. 

LDA is a matrix decomposition method, and I would not recommend to link it with clustering so directly.

But it could in principle go to the bluster package. Not sure if they would approve but it could be asked by opening an issue. That might take more time, however.. therefore I would meanwhile proceed with implementing these in mia, and later we can move them elsewhere if necessary? LDA and NMF have clear uses, it would be useful to have them easily accessible.

TuomasBorman commented:
With these explanations, I am fine with mia implementation and reducedDim()

Commit:
add comments to getLDA and addLDA

theotimeee committed:
Merge branch 'devel' into feature_loadings_methods

Commit:
add input checks

Commit:
Merge branch 'feature_loadings_methods' of https://github.com/microbiome/mia into feature_loadings_methods

Commit:
add tests for addLDA

TuomasBorman commented on R/addLDA.R:
```diff
@@ -0,0 +1,105 @@
+#' @param x a \code{\link[TreeSummarizedExperiment:TreeSummarizedExperiment-class]{TreeSummarizedExperiment}}
+#'  object.
+#'   
+#' @param k Integer, number of latent vectors/topics 
+#'   (By default: \code{2}).
```
We are trying to harmonize these descriptions little bit. This could be:

Integer scalar. A number of latent vectors/topics. (Default: \code{2})

The type. Detailed description. Default value.

TuomasBorman commented on R/addLDA.R:
```diff
@@ -0,0 +1,105 @@
+#' @param k Integer, number of latent vectors/topics 
+#'   (By default: \code{2}).
+#' 
+#' @param name The name to be used to store the result in the reducedDims of the
+#'  output (By default: \code{name = "LDA"}).
```
This could be:

Character scalar. The name to be used to store the result in the reducedDims of the output. (Default: \code{"LDA"})

TuomasBorman commented on R/addLDA.R:
```diff
@@ -0,0 +1,105 @@
+#' @param name The name to be used to store the result in the reducedDims of the
+#'  output (By default: \code{name = "LDA"}).
+#'  
+#' @param assay.type a single \code{character} value for specifying which
+#'   assay to use for LDA ordination.
+#' 
```
Character scalar. Specifies which assay to use for LDA ordination. (Default: \code{"counts"})

TuomasBorman commented on R/addLDA.R:
```diff
@@ -0,0 +1,105 @@
+#' @export
+#' @rdname addLDA
+setMethod("getLDA", "SummarizedExperiment",
+    function(x, k=2, assay.type = "counts", ...){
```
k = 2 instead of k=2 --> use similar coding-style (you have assay.type = "counts" already)

TuomasBorman commented on R/utils.R:
```diff
@@ -349,6 +349,38 @@
+            call. = FALSE)
+    }
+    # Throw warning if values of reducedDim are overwritten
+    if ( c(name) %in% names(reducedDims(x)) ){
```
c(name) --> name

TuomasBorman commented on R/utils.R:
```diff
@@ -349,6 +349,38 @@
+        warning(
+          "The following values are already present in `reducedDims` and will ",
+          "be overwritten: '",
+          paste(name),
```
No need to use paste(), it should work without it (moreover, Bioc guidelines gives say that they should not be used inside warning/stop)

TuomasBorman reviewed (changes requested):
Minor things

Commit:
review modifications

TuomasBorman commented on tests/testthat/test-5addLDA.R:
```diff
@@ -0,0 +1,40 @@
+  )
+  expect_error(
+    addLDA(GlobalPatterns, k = 2, assay.type = "counts", name = TRUE)
+  )
```
Can you still add test that the calculated result is same with getLDA, addLDA and topicmodels::LDA?

theotimeee replied:
I added tests but there is an average diff of 7*10**(-5) that comes from adding the result to reducedDims I think, so I put a tolerance level of 10**(-4)

TuomasBorman commented on R/addLDA.R:
```diff
@@ -0,0 +1,105 @@
+#' @name addLDA
+#' 
+#' @examples
+#' data(GlobalPatterns)
+#' 
+#' # Reduce the number of features 
+#' tse <- agglomerateByPrevalence(GlobalPatterns, rank="Phylum", prevalence=0.99, update.tree = TRUE)
+#' 
```
tse <- GlobalPatterns

And then run aglomeratebyPrevalence

TuomasBorman replied:
btw does

agglomerateByPrevalence(tse, rank="Phylum")

That would be more simple

TuomasBorman commented on R/addLDA.R:
```diff
@@ -0,0 +1,105 @@
+        }
+        df <- as.data.frame(t(assay(x, assay.type)))
+        # Estimate LDA model using VEM algorithm
+        lda_model <- topicmodels::LDA(df, k)
```
Pass ... so that user can choose the method and control parameters of topicmodels::LDA

TuomasBorman commented on R/addLDA.R:
```diff
@@ -0,0 +1,105 @@
+#' @param assay.type Character scalar. Specifies which assay to use for LDA 
+#'  ordination. (Default: \code{"counts"})
+#' 
+#' @param ... optional arguments.
```
In details section say clearly that topicsmodel::LDA is used

Here say that additional parameters are passed to that function

TuomasBorman commented on R/addLDA.R:
```diff
@@ -0,0 +1,105 @@
+        scores <- t(as.data.frame(posteriors$topics))
+        loadings <- t(as.data.frame(posteriors$terms)) 
+        # Add loadings as attribute of the scores matrix
+        attr(scores, "loadings") <- loadings
```
You could also add the model to attributes

TuomasBorman commented:
Usually, the number of topics "k" is determined by iterating over several numbers. The best fit is evaluated with coherence and perplexity metrics.

Should k be a vector of values so that the function iterates over them calculating LDA for all different number of topics? The function could evaluate the goodness of fit and then return the best model. The function could also return a table which tells the metrics for each k value (that could be then easily plotted). 

@antagomir

antagomir commented:
Ok to me to add that. 

However, if it feels a lot to add now, I would focus on finalizing the present one (with a single fixed k) first, and then adding that multi-k optimization as a separate PR.

TuomasBorman commented:
That would require something like this


```
addLDA <- function(eval.metric = "perplexity"){
    if(eval.metric == "coherence" && !require("topicdoc") ){
        stop("'topicmodel' package must be installed in order to calculate coherence.")
    }
    models <- lapply(k, function(x){
        LDA(k = x)
    }
    
    metric <- .calc_metrics(models)
    
    # Get the model with lowest eval.metric (dfault value perplexity).
    
    # Get scores
    # Add loadings, model, and metrics data.frame to attributes

}

.calc_metrics <- function(x){
    # Loop through models
    # Calculate perplexity with topicmodels
    # Calculate coherence (I believe you have to use topicdoc package). Calculate it if that package is available
    # Create df from both metrics
}
```

Commit:
review modifs

Commit:
call agglomerateByPrevalence without prevalence level in examples

Commit:
merge devel

Merge branch 'devel' of https://github.com/microbiome/mia into feature_loadings_methods

# Conflicts:
#	NEWS

theotimeee committed:
Merge branch 'devel' into feature_loadings_methods

TuomasBorman commented on R/utils.R:
```diff
@@ -349,6 +349,38 @@
+        stop("'values' must be a matrix.", call. = FALSE)
+    }
+    # Check names
+    if( !is.character(name) && length(name) > 1L ){
```
Use .is_a_string()

TuomasBorman commented on R/addLDA.R:
```diff
@@ -0,0 +1,113 @@
+#' @param x a \code{\link[TreeSummarizedExperiment:TreeSummarizedExperiment-class]{TreeSummarizedExperiment}}
+#'  object.
+#'   
+#' @param k Integer scalar. A number of latent vectors/topics. 
```
Looks good. Can you still put the class inside  \code{] to make it nicer. --> \code{Integer scalar} (Check other also)

TuomasBorman commented on R/addLDA.R:
```diff
@@ -0,0 +1,113 @@
+        if( !.is_an_integer(k) ){
+            stop("'k' must be an integer.", call. = FALSE)
+        }
+        if( !.is_a_string(assay.type) ){
+            stop("'assay.type' must be a non-empty single character value.",
+              call. = FALSE)
+        }
```
Intead, use .check_assay_present which also checks that the assay is available

TuomasBorman commented:
NMF would go with this same template I think. However, there k (rank) can be numeric vector and the function finds optimal k by itself

TuomasBorman commented:
For LDA we could have another PR for multiple k values, if you are not able to implement it now

The idea was this:


```
data("GlobalPatterns")
tse <- GlobalPatterns
tse <- agglomerateByPrevalence(tse, rank="Phylum")

eval.metric <- "perplexity"
k <- c(2, 3, 4, 5, 6, 7, 8, 9, 10)
k <- 5
assay.type <- "counts"

mat <- assay(tse, assay.type)

df <- as.data.frame(t(mat))




if( eval.metric == "coherence" ){
    .require_package("topicdoc")
}

models <- lapply(k, function(i){
    topicmodels::LDA(df, i)
})

metrics <- .calculate_lda_metrics(models, k)

lda_model <- models[[ which.max(metrics[[eval.metric]]) ]]

posteriors <- topicmodels::posterior(lda_model, df)
scores <- t(as.data.frame(posteriors$topics))
attr(scores, "loadings") <- loadings
attr(scores, "model") <- lda_model
attr(scores, "eval_metric") <- metrics 

.calculate_lda_metrics <- function(models, k){
    metrics <- lapply(models, function(model){
        res <- topicmodels::perplexity(model)
        names(res) <- "perplexity"
        if( require("topicdoc") ){
            coherence <- topicdoc::topic_coherence(model, df)
            names(coherence) <- paste0("coherence_", seq_len(length(coherence)))
            coherence <- c(coherence, coherence_mean = mean(coherence))
            res <- c(res, coherence)
        }
        return(res)
    })
    metrics <- as.data.frame( dplyr::bind_rows(metrics) )
    metrics[["k"]] <- k
    return(metrics)
}

#################################
df <- attr(scores, "eval_metric")

library(ggplot2)

ggplot(df) +
    geom_point(aes(x = k, y = perplexity), color = "red")
ggplot(df) +
    geom_point(aes(x = k, y = mean_coherence), color = "blue")


```

Commit:
address review comments

TuomasBorman committed:
Update addLDA.R

theotimeee committed:
Merge branch 'devel' into feature_loadings_methods

TuomasBorman commented on R/utils.R:
```diff
@@ -349,6 +349,38 @@
+    # Throw warning if values of reducedDim are overwritten
+    if ( name %in% names(reducedDims(x)) ){
+        warning(
+          "The following values are already present in `reducedDims` and will ",
+          "be overwritten: '",
+          name,
+          "'. Consider using the 'name' argument to specify alternative ",
+          "names.", call. = FALSE)
+    }
```
Indentation is off

theotimeee replied:
I fixed the indentation and modified the input check to allow dist objects as well as matrices.

TuomasBorman reviewed (changes requested):
Is this PR ready? If the modification on multiple k values do not come to this PR, I think this is good to go

Commit:
modify input check and indentation in .add_values_to_reducedDims

Commit:
Merge branch 'feature_loadings_methods' of https://github.com/microbiome/mia into feature_loadings_methods

TuomasBorman committed:
up

TuomasBorman committed:
up

TuomasBorman committed:
up

TuomasBorman committed:
up

TuomasBorman committed:
up

TuomasBorman committed:
up

TuomasBorman committed:
up

TuomasBorman committed:
up

TuomasBorman committed:
Merge branch 'devel' into feature_loadings_methods

TuomasBorman committed:
up

TuomasBorman committed:
up
